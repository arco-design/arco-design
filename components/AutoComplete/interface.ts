import { CSSProperties, ReactNode, ReactElement } from 'react';
import { InputProps } from '../Input';
import { OptionInfo, SelectProps } from '../Select/interface';

import { AvailableVirtualListProps } from '../_class/VirtualList';

type PartialSelectProps = Pick<
  SelectProps,
  | 'filterOption'
  | 'dropdownRender'
  | 'triggerProps'
  | 'getPopupContainer'
  | 'defaultActiveFirstOption'
>;

/**
 * @title AutoComplete
 */
export interface AutoCompleteProps extends PartialSelectProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 是否禁用
   * @en Whether to disable the autocomplete
   */
  disabled?: boolean;
  /**
   * @zh 自定义输入框/数据源
   * @en Custom input/source data
   */
  children?: ReactNode;
  /**
   * @zh 是否允许一键清除
   * @en Whether to allow clear the content
   */
  allowClear?: boolean;
  /**
   * @zh 自动完成的数据源
   * @en Data source
   */
  data?: (string | { value: string; name: string; [key: string]: any } | ReactNode)[];
  /**
   * @zh 输入框提示
   * @en Placeholder of input
   */
  placeholder?: string;
  /**
   * @zh 自动补全组件的默认值
   * @en The default value of the autocomplete.
   */
  defaultValue?: string;
  /**
   * @zh 自动补全组件的值（受控模式）
   * @en The value of the autocomplete.
   */
  value?: string;
  /**
   * @zh 是否是错误状态。(废弃，下个大版本移除，使用 status='error' 替代)
   * @en Whether the textarea is error.(Deprecated, removed in the next major version, use status='error' instead)
   * @deprecated use status="error"
   */
  error?: boolean;
  /**
   * @zh 状态
   * @en Status
   * @version 2.45.0
   */
  status?: 'error' | 'warning';
  /**
   * @zh `strict: true` 的时候大小写敏感
   * @en Case sensitive when set `strict: true`
   */
  strict?: boolean;
  /**
   * @zh 是否处于加载状态。
   * @en Whether the component is loading data.
   * @version 2.10.0
   */
  loading?: boolean;
  /**
   * @zh 自定义触发元素
   * @en Custom trigger element
   * @defaultValue <Input />
   * @version `() => ReactElement` in 2.31.0
   */
  triggerElement?: ReactElement | (({ value }) => ReactElement);
  /**
   * @zh 搜索补全时的回调
   * @en Callback when searching items
   */
  onSearch?: (value: string) => void;
  /**
   * @zh 点击补全项时的回调
   * @en Callback when an option is selected.
   */
  onSelect?: (value: string, option: OptionInfo) => void;
  /**
   * @zh 输入或者点击补全项，value 改变时的回调（仅在点击补全项时存在第二个参数）
   * @en Callback when an option is selected or input value changes
   */
  onChange?: (value: string, option?: OptionInfo) => void;
  /**
   * @zh 按下回车键的回调
   * @en Callback when Enter is pressed
   * @version `activeOption` in 2.25.1
   */
  onPressEnter?: (event, activeOption?: OptionInfo) => void;
  /**
   * @zh 聚焦时的回调
   * @en Callback when component gets focus
   */
  onFocus?: (event) => void;
  /**
   * @zh 失去焦点的回调
   * @en Callback when component is blurred
   */
  onBlur?: (event) => void;
  /**
   * @zh 传递虚拟滚动属性。
   * @en Virtual scroll properties.
   * @version 2.2.0
   */
  virtualListProps?: AvailableVirtualListProps;
  /**
   * @zh 传递 Input 组件的属性。
   * @en Properties of Input component.
   * @version 2.10.0
   */
  inputProps?: Partial<InputProps>;
}
