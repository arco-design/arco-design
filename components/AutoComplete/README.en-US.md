`````
Component / Data Entry

# AutoComplete

An input box or custom input control, with enhanced automatic completion functionality.
`````

%%Content%%

## API

### AutoComplete

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|-|
|className|Additional css class|`string \| string[]`|`-`|-|
|disabled|Whether to disable the autocomplete|`boolean`|`-`|-|
|children|Custom input/source data|`React.ReactNode`|`-`|-|
|allowClear|Whether to allow clear the content|`boolean`|`-`|-|
|data|Data source|`(string \| { value: string; name: string; [key: string]: any } \| React.ReactNode)[]`|`-`|-|
|placeholder|Placeholder of input|`string`|`-`|-|
|defaultValue|The default value of the autocomplete.|`string`|`-`|-|
|value|The value of the autocomplete.|`string`|`-`|-|
|error|Error style|`boolean`|`-`|-|
|strict|Case sensitive when set `strict: true`|`boolean`|`-`|-|
|loading|Whether the component is loading data.|`boolean`|`-`|2.10.0|
|triggerElement|Custom trigger element|`ReactElement \| (({ value }) => ReactElement)`|`<Input />`|`() => ReactElement` in 2.31.0|
|onSearch|Callback when searching items|`(value: string) => void`|`-`|-|
|onSelect|Callback when an option is selected.|`(value: string, option: OptionInfo) => void`|`-`|-|
|onChange|Callback when an option is selected or input value changes|`(value: string, option?: OptionInfo) => void`|`-`|-|
|onPressEnter|Callback when Enter is pressed|`(event, activeOption?: OptionInfo) => void`|`-`|`activeOption` in 2.25.1|
|onFocus|Callback when component gets focus|`(event) => void`|`-`|-|
|onBlur|Callback when component is blurred|`(event) => void`|`-`|-|
|virtualListProps|Virtual scroll properties.|`AvailableVirtualListProps`|`-`|2.2.0|
|inputProps|Properties of Input component.|`Partial<InputProps>`|`-`|2.10.0|
|filterOption|If it's true, filter options by input value. If it's a function, filter options base on the function.|`boolean \| ((inputValue: string, option: ReactElement) => boolean)`|`true`|-|
|dropdownRender|Customize dropdown content|`(menu: ReactNode) => ReactNode`|`-`|-|
|triggerProps|Pass all `Trigger` component properties|`Partial<TriggerProps>`|`-`|-|
|getPopupContainer|To set the container of the dropdown.|`(node: HTMLElement) => Element`|`-`|-|
|defaultActiveFirstOption|Whether to highlight the first option by default|`boolean`|`true`|-|

### AutoComplete.Option

Same as `Select.Option`, refer to [Select](/react/en-US/components/select).

### AutoComplete.OptGroup

Same as `Select.OptGroup`, refer to [Select](/react/en-US/components/select).
