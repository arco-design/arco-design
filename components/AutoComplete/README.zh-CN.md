`````
组件 / 数据输入

# 自动补全 AutoComplete

输入框或自定义输入控件的自动补全功能。
`````

%%Content%%

## API

### AutoComplete

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|style|节点样式|`CSSProperties`|`-`|-|
|className|节点类名|`string \| string[]`|`-`|-|
|disabled|是否禁用|`boolean`|`-`|-|
|children|自定义输入框/数据源|`React.ReactNode`|`-`|-|
|allowClear|是否允许一键清除|`boolean`|`-`|-|
|data|自动完成的数据源|`(string \| { value: string; name: string; [key: string]: any } \| React.ReactNode)[]`|`-`|-|
|placeholder|输入框提示|`string`|`-`|-|
|defaultValue|自动补全组件的默认值|`string`|`-`|-|
|value|自动补全组件的值（受控模式）|`string`|`-`|-|
|error|报错状态|`boolean`|`-`|-|
|strict|`strict: true` 的时候大小写敏感|`boolean`|`-`|-|
|loading|是否处于加载状态。|`boolean`|`-`|2.10.0|
|triggerElement|自定义触发元素|`ReactElement \| (({ value }) => ReactElement)`|`<Input />`|`() => ReactElement` in 2.31.0|
|onSearch|搜索补全时的回调|`(value: string) => void`|`-`|-|
|onSelect|点击补全项时的回调|`(value: string, option: OptionInfo) => void`|`-`|-|
|onChange|输入或者点击补全项，value 改变时的回调（仅在点击补全项时存在第二个参数）|`(value: string, option?: OptionInfo) => void`|`-`|-|
|onPressEnter|按下回车键的回调|`(event, activeOption?: OptionInfo) => void`|`-`|`activeOption` in 2.25.1|
|onFocus|聚焦时的回调|`(event) => void`|`-`|-|
|onBlur|失去焦点的回调|`(event) => void`|`-`|-|
|virtualListProps|传递虚拟滚动属性。|`AvailableVirtualListProps`|`-`|2.2.0|
|inputProps|传递 Input 组件的属性。|`Partial<InputProps>`|`-`|2.10.0|
|filterOption|是否根据输入的值筛选数据。如果传入函数的话，接收 `inputValue` 和 `option` 两个参数，当option符合筛选条件时，返回 `true`，反之返回 `false`。|`boolean \| ((inputValue: string, option: ReactElement) => boolean)`|`true`|-|
|dropdownRender|自定义弹出内容。|`(menu: ReactNode) => ReactNode`|`-`|-|
|triggerProps|可以接受所有 `Trigger` 的 `Props`|`Partial<TriggerProps>`|`-`|-|
|getPopupContainer|弹出框挂载的父节点。|`(node: HTMLElement) => Element`|`-`|-|
|defaultActiveFirstOption|是否默认高亮第一个选项|`boolean`|`true`|-|

### AutoComplete.Option

同 `Select.Option`, 参考 [Select 文档](/react/components/select) 。

### AutoComplete.OptGroup

同 `Select.OptGroup`, 参考 [Select 文档](/react/components/select) 。
