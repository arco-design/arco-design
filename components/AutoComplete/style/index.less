@import './token.less';

@auto-complete-prefix-cls: ~'@{prefix}-autocomplete';
@select-prefix-cls: ~'@{prefix}-select';

.@{auto-complete-prefix-cls} {
  &-popup .@{select-prefix-cls}-popup {
    border: 1px solid @auto-complete-popup-color-border;
    border-radius: @auto-complete-popup-border-radius;
    background-color: var(~'@{arco-cssvars-prefix}-color-bg-popup');
    box-shadow: @auto-complete-popup-box-shadow;

    .@{select-prefix-cls}-popup-inner {
      padding: @auto-complete-popup-padding-vertical 0;
      max-height: @auto-complete-popup-max-height;
    }

    .@{select-prefix-cls}-option {
      height: @auto-complete-option-height;
      padding: 0 @auto-complete-option-padding-horizontal;
      font-size: @auto-complete-popup-font-size;
      line-height: $height;

      .option-color(@status) {
        @class-suffix: if(@status = default, ~'', ~'-@{status}');

        &@{class-suffix} {
          color: ~'@{auto-complete-option-color-text_@{status}}';
          background-color: ~'@{auto-complete-option-color-bg_@{status}}';
        }
      }

      .option-color(default);
      .option-color(selected);
      .option-color(hover);
      // disabled 优先级最高，放在最后
      .option-color(disabled);

      &-selected {
        font-weight: @auto-complete-option-font-weight_selected;
      }
    }
  }
}
