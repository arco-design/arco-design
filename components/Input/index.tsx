import Input from './input';

import type { InputProps, TextAreaProps } from './interface';
import type { InputPasswordProps } from './password';
import type { InputSearchProps } from './search';

export { InputProps, TextAreaProps, InputPasswordProps, InputSearchProps };

export default Input;
