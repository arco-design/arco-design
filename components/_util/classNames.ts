import { isArray, isObject, isString } from './is';
import warning from './warning';

type ClassNamesArg = string | string[] | { [key: string]: any } | undefined | null | boolean;

export default function (...args: ClassNamesArg[]): string {
  const length = args.length;
  let classNames: string[] = [];
  for (let i = 0; i < length; i++) {
    const v = args[i];
    if (!v) {
      continue;
    }
    if (isString(v)) {
      classNames.push(v);
    } else if (isArray(v)) {
      classNames = classNames.concat(v);
    } else if (isObject(v)) {
      Object.keys(v).forEach((k) => {
        if (v[k]) {
          classNames.push(k);
        }
      });
    } else {
      warning(true, 'arguments must be one of string/array/object.');
    }
  }
  return [...new Set(classNames)].join(' ');
}
