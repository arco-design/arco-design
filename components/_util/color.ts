// https://github.com/scttcper/tinycolor
export const hsvToRgb = (h: number, s: number, v: number) => {
  const i = Math.floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - f * s);
  const t = v * (1 - (1 - f) * s);
  const mod = i % 6;
  const r = [v, q, p, p, t, v][mod];
  const g = [t, v, v, q, p, p][mod];
  const b = [p, p, t, v, v, q][mod];

  return {
    r: Math.round(r * 255),
    g: Math.round(g * 255),
    b: Math.round(b * 255),
  };
};

export const rgbToHsv = (r: number, g: number, b: number) => {
  r /= 255;
  g /= 255;
  b /= 255;

  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  let h = 0;
  const v = max;
  const d = max - min;
  const s = max === 0 ? 0 : d / max;

  if (max === min) {
    h = 0;
  } else {
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0);
        break;
      case g:
        h = (b - r) / d + 2;
        break;
      case b:
        h = (r - g) / d + 4;
        break;
      default:
        break;
    }

    h /= 6;
  }

  return { h, s, v };
};

// <http://www.w3.org/TR/css3-values/#integers>
const CSS_INTEGER = '[-\\+]?\\d+%?';

// <http://www.w3.org/TR/css3-values/#number-value>
const CSS_NUMBER = '[-\\+]?\\d*\\.\\d+%?';

// Allow positive/negative integer/number.  Don't capture the either/or, just the entire outcome.
const CSS_UNIT = `(?:${CSS_NUMBER})|(?:${CSS_INTEGER})`;

// Actual matching.
// Parentheses and commas are optional, but not required.
// Whitespace can take the place of commas or opening paren
const PERMISSIVE_MATCH3 = `[\\s|\\(]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})\\s*\\)?`;
const PERMISSIVE_MATCH4 = `[\\s|\\(]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})[,|\\s]+(${CSS_UNIT})\\s*\\)?`;

const matchers = {
  rgb: new RegExp('rgb' + PERMISSIVE_MATCH3),
  rgba: new RegExp('rgba' + PERMISSIVE_MATCH4),
  hex3: /^#?([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex6: /^#?([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/,
  hex4: /^#?([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex8: /^#?([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/,
};

export function parseIntFromHex(val: string): number {
  return parseInt(val, 16);
}

export function convertHexToDecimal(h: string): number {
  return parseIntFromHex(h) / 255;
}

export function formatInputToRgb(color: string): { r: number, g: number, b: number, a?: number } | false {
  let match = matchers.rgb.exec(color);
  if (match) {
    return { r: parseInt(match[1]), g: parseInt(match[2]), b: parseInt(match[3]) };
  }

  match = matchers.rgba.exec(color);
  if (match) {
    return { r: parseInt(match[1]), g: parseInt(match[2]), b: parseInt(match[3]), a: parseFloat(match[4]) };
  }

  match = matchers.hex8.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1]),
      g: parseIntFromHex(match[2]),
      b: parseIntFromHex(match[3]),
      a: convertHexToDecimal(match[4]),
    };
  }

  match = matchers.hex6.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1]),
      g: parseIntFromHex(match[2]),
      b: parseIntFromHex(match[3]),
    };
  }

  match = matchers.hex4.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1] + match[1]),
      g: parseIntFromHex(match[2] + match[2]),
      b: parseIntFromHex(match[3] + match[3]),
      a: convertHexToDecimal(match[4] + match[4]),
    };
  }

  match = matchers.hex3.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1] + match[1]),
      g: parseIntFromHex(match[2] + match[2]),
      b: parseIntFromHex(match[3] + match[3]),
    };
  }

  return false;
}

export function hexToRgb(color: string): any {
  color = color.trim().toLowerCase();
  if (color.length === 0) {
    return false;
  }

  let match = matchers.hex6.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1]),
      g: parseIntFromHex(match[2]),
      b: parseIntFromHex(match[3]),
    };
  }

  match = matchers.hex3.exec(color);
  if (match) {
    return {
      r: parseIntFromHex(match[1] + match[1]),
      g: parseIntFromHex(match[2] + match[2]),
      b: parseIntFromHex(match[3] + match[3]),
    };
  }

  return false;
}

export const rgbToHex = (r: number, g: number, b: number) => {
  const hex = [
    Math.round(r).toString(16).padStart(2, '0'),
    Math.round(g).toString(16).padStart(2, '0'),
    Math.round(b).toString(16).padStart(2, '0'),
  ];

  return hex.join('').toUpperCase();
};

export const rgbaToHex = (r: number, g: number, b: number, a: number) => {
  const hex = [
    Math.round(r).toString(16).padStart(2, '0'),
    Math.round(g).toString(16).padStart(2, '0'),
    Math.round(b).toString(16).padStart(2, '0'),
    Math.round(a * 255).toString(16).padStart(2, '0'),
  ];

  return hex.join('').toUpperCase();
};
