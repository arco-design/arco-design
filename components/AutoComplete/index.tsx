import React, { useContext, useRef, isValidElement, useState, useImperativeHandle } from 'react';
import cs from '../_util/classNames';
import Input from '../Input';
import { ConfigContext } from '../ConfigProvider';
import useMergeValue from '../_util/hooks/useMergeValue';
import Select, { SelectHandle } from '../Select/select';
import { OptionInfo, SelectProps } from '../Select/interface';
import { isSelectOption, isSelectOptGroup } from '../Select/utils';
import { Enter, Esc } from '../_util/keycode';
import omit from '../_util/omit';
import { RefInputType } from '../Input/interface';
import IconLoading from '../../icon/react-icon/IconLoading';
import { AutoCompleteProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const IMPOSSIBLE_VALUE = `Autocomplete_${Math.random()}`;

const { Option } = Select;

const defaultProps: AutoCompleteProps = {
  defaultActiveFirstOption: true,
  triggerElement: <Input />,
};

function AutoComplete(baseProps: AutoCompleteProps, ref) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<AutoCompleteProps>(
    baseProps,
    defaultProps,
    componentConfig?.AutoComplete
  );
  const {
    style,
    className,
    children,
    data,
    defaultValue,
    value: propValue,
    placeholder,
    error,
    disabled,
    strict,
    allowClear,
    loading,
    defaultActiveFirstOption,
    triggerElement,
    getPopupContainer,
    dropdownRender,
    virtualListProps,
    onFocus,
    onBlur,
    onChange,
    onSearch,
    onSelect,
    onPressEnter,
    inputProps,
  } = props;

  const [value, setValue] = useMergeValue('', {
    defaultValue,
    value: propValue,
  });
  const [isFocused, setIsFocused] = useState(false);

  const refInput = useRef(null);
  const refSelect = useRef<SelectHandle>(null);

  const prefixCls = getPrefixCls('autocomplete');
  const filterOption =
    'filterOption' in props
      ? props.filterOption
      : (inputValue, option) => {
          if (strict) {
            return option.props.value.indexOf(inputValue) > -1;
          }
          return option.props.value.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
        };

  const childNodes = React.Children.toArray(children);
  let selectChildren = null;

  if (childNodes.length && (isSelectOption(childNodes[0]) || isSelectOptGroup(childNodes[0]))) {
    selectChildren = children;
  } else if (data && data.length) {
    selectChildren = data.map((item, index) => {
      if (isValidElement(item)) {
        return item;
      }
      if (typeof item === 'string') {
        return (
          <Option key={index} value={item}>
            {item}
          </Option>
        );
      }
      if (typeof item === 'object') {
        const { value, name } = item as any;
        return (
          <Option key={index} value={value} extra={omit(item, ['value', 'name'])}>
            {name}
          </Option>
        );
      }
      return null;
    });
  }

  useImperativeHandle(ref, () => refInput.current);

  const TriggerElement = React.cloneElement(triggerElement, {
    ref: (node) => {
      refInput.current = node;

      const { ref: originRef } = triggerElement as any;
      if (typeof originRef === 'function') {
        originRef(node);
      }
    },
    className: cs(`${prefixCls}`, inputProps && inputProps.className, className),
    style,
    value,
    placeholder,
    error,
    disabled,
    allowClear,
    ...inputProps,
    // Empty tag to ensure the consistency of the dom structure of input, such that input won't accidentally lose focus due to structure change on input.
    suffix: loading ? <IconLoading /> : inputProps?.suffix || <i />,
    onFocus: (event) => {
      setIsFocused(true);
      onFocus && onFocus(event);
      inputProps && inputProps.onFocus && inputProps.onFocus(event);
    },
    onBlur: (event) => {
      setIsFocused(false);
      onBlur && onBlur(event);
      inputProps && inputProps.onBlur && inputProps.onBlur(event);
    },
    onKeyDown: (event) => {
      const keyCode = event.keyCode || event.which;
      refSelect.current && refSelect.current.hotkeyHandler(event);

      if (keyCode === Enter.code && onPressEnter) {
        let activeOption;
        if (refSelect.current) {
          activeOption = refSelect.current.getOptionInfoByValue(
            refSelect.current.activeOptionValue
          );
        }
        onPressEnter(event, activeOption);
      }

      if (keyCode === Esc.code) {
        refInput.current && refInput.current.blur && refInput.current.blur();
      }

      inputProps && inputProps.onKeyDown && inputProps.onKeyDown(event);
    },
    onChange: (value, event) => {
      setValue(value);
      onSearch && onSearch(value);
      onChange && onChange(value);
      inputProps && inputProps.onChange && inputProps.onChange(value, event);
    },
  });

  const triggerProps: SelectProps['triggerProps'] = {
    popupVisible: !!(isFocused && (data?.length || React.Children.count(children))),
    ...props.triggerProps,
    // Other trigger types are not supported yet
    trigger: 'focus',
    className: [`${prefixCls}-popup`].concat(props.triggerProps && props.triggerProps.className),
  };

  const selectProps: Partial<SelectProps> = {
    triggerElement: TriggerElement,
    // Guarantee that onChange can always be triggered
    value: IMPOSSIBLE_VALUE,
    inputValue: value,
    defaultActiveFirstOption,
    triggerProps,
    getPopupContainer,
    dropdownRender,
    filterOption,
    virtualListProps,
    notFoundContent: null,
    onChange: (value: string, option) => {
      setValue(value);
      onChange && onChange(value, option as OptionInfo);
      value && onSelect && onSelect(value, option as OptionInfo);
      // Blur the input on option change
      refInput.current && refInput.current.blur && refInput.current.blur();
    },
  };

  return (
    <Select ref={refSelect} {...selectProps}>
      {selectChildren}
    </Select>
  );
}

const ForwardRefAutoComplete = React.forwardRef<RefInputType, AutoCompleteProps>(AutoComplete);

const AutoCompleteComponent = ForwardRefAutoComplete as typeof ForwardRefAutoComplete & {
  Option: typeof Select.Option;
  OptGroup: typeof Select.OptGroup;
};

AutoCompleteComponent.displayName = 'AutoComplete';

AutoCompleteComponent.Option = Select.Option;
AutoCompleteComponent.OptGroup = Select.OptGroup;

export default AutoCompleteComponent;

export { AutoCompleteProps };
