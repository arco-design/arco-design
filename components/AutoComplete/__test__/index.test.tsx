import React from 'react';
import { mount } from 'enzyme';
import { ArrowDown, ArrowUp, Enter, Esc } from '../../_util/keycode';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import AutoComplete from '..';
import Input from '../../Input';

const { TextArea } = Input;

mountTest(AutoComplete);
componentConfigTest(AutoComplete, 'AutoComplete');

function mountAutoComplete(component: React.ReactElement) {
  return mount(component);
}

afterEach(() => {
  jest.clearAllMocks();
});

describe('AutoComplete', () => {
  it('AutoComplete simple use correctly', () => {
    const wrapper = mountAutoComplete(
      <AutoComplete
        placeholder="请输入城市"
        data={[
          '北京',
          '上海',
          '深圳',
          {
            name: '杭州',
            value: '杭州',
          },
        ]}
        triggerElement={<TextArea style={{ width: 520, height: 96 }} />}
      />
    );
    wrapper.find('TextArea').simulate('focus');
    expect(wrapper.find('.arco-trigger')).toHaveLength(1);
  });

  it('AutoComplete with children', () => {
    const wrapper = mountAutoComplete(
      <AutoComplete>
        <AutoComplete.Option value="beijing">Beijing</AutoComplete.Option>
        <AutoComplete.Option value="shanghai">Shanghai</AutoComplete.Option>
      </AutoComplete>
    );
    wrapper.find('input').simulate('focus');
    expect(wrapper.find('.arco-trigger')).toHaveLength(1);
  });

  it('AutoComplete callback correctly', () => {
    const onSearch = jest.fn();
    const onSelect = jest.fn();
    const onChange = jest.fn();
    const onFocus = jest.fn();
    const onKeyDown = jest.fn();
    const onPressEnter = jest.fn();
    const onInputChange = jest.fn();

    const wrapper = mountAutoComplete(
      <AutoComplete
        placeholder="请输入城市"
        data={['北京', '上海', '深圳', '杭州']}
        style={{ width: 260, marginRight: 20, marginBottom: 20 }}
        onSearch={onSearch}
        onSelect={onSelect}
        onChange={onChange}
        onFocus={onFocus}
        inputProps={{
          onKeyDown,
          onChange: onInputChange,
        }}
        onPressEnter={onPressEnter}
      />
    );
    wrapper.find('input').simulate('focus');
    expect(onFocus).toBeCalled();

    wrapper.find('input').simulate('change', {
      target: {
        value: '上',
      },
    });
    expect(onSearch).toBeCalled();
    expect(onChange).toBeCalled();
    expect(onInputChange).toBeCalled();

    wrapper.find('input').simulate('keydown', {
      keyCode: Enter.code,
    });
    expect(onPressEnter).toBeCalled();

    wrapper.find('.arco-select-option').at(0).simulate('click');
    expect(onSelect).toBeCalled();
  });

  it('keyboard events correctly', () => {
    const wrapper = mountAutoComplete(
      <AutoComplete
        placeholder="请输入城市"
        data={['北京', '上海', '深圳', '杭州']}
        style={{ width: 260, marginRight: 20, marginBottom: 20 }}
      />
    );
    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('keyDown', { keyCode: ArrowDown.code });
    expect(wrapper.find('.arco-select-option-hover').text()).toBe('上海');

    wrapper.find('input').simulate('keyDown', { keyCode: ArrowUp.code });
    expect(wrapper.find('.arco-select-option-hover').text()).toBe('北京');

    wrapper.find('input').simulate('keyDown', { keyCode: Enter.code });
    wrapper.find('input').simulate('keyDown', { keyCode: Esc.code });
    expect(wrapper.find('input').prop('value')).toBe('北京');
  });
});
