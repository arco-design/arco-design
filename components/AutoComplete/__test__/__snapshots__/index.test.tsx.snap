// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.AutoComplete default 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-autocomplete"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        class="arco-input arco-input-size-default"
        tabindex="0"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <i />
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.AutoComplete set className globally 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-autocomplete global-default-class"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        class="arco-input arco-input-size-default"
        tabindex="0"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <i />
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.AutoComplete set className globally and component set className itself. 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-autocomplete component-class"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        class="arco-input arco-input-size-default"
        tabindex="0"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <i />
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.AutoComplete set data-* property 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-autocomplete"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        class="arco-input arco-input-size-default"
        data-test-name="AutoComplete"
        tabindex="0"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <i />
      </span>
    </span>
  </span>
</div>
`;
